### PR TITLE
Replace `std::regex` by re2

### PR DESCRIPTION
**Pull request by rschu1ze**
std::regex is [generally disliked](https://www.reddit.com/r/cpp_questions/comments/s2d6p9/does_anyone_here_use_regex_very_much_is_it/) ...

~~Adresses~~
- ~~https://github.com/ClickHouse/ClickHouse/issues/57931~~
- ~~https://github.com/ClickHouse/support-escalation/issues/1358~~

... but this PR is still nice.

### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Usages of `std::regex` (`#include <regex>`) are replaced by re2. New usages are disallowed.

### CODE REVIEW

**Comment by robot-clickhouse-ci-1** [comment]
*This is an automated comment for commit 8d4b519bb1aebb40bcbc45226e99af3beff154cd with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/58458/8d4b519bb1aebb40bcbc45226e99af3beff154cd/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/58458/8d4b519bb1aebb40bcbc45226e99af3beff154cd/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

File: src/Server/MySQLHandler.cpp [hunk 1]
```diff
@@ -530,16 +538,15 @@ void MySQLHandlerSSL::finishHandshakeSSL(
...
-    static const std::regex expr{
+    static const re2::RE2 expr(
         "(^(SET NAMES(.*)))"
         "|(^(SET character_set_results(.*)))"
         "|(^(SET FOREIGN_KEY_CHECKS(.*)))"
         "|(^(SET AUTOCOMMIT(.*)))"
         "|(^(SET sql_mode(.*)))"
         "|(^(SET @@(.*)))"
-        "|(^(SET SESSION TRANSACTION ISOLATION LEVEL(.*)))"
-        , std::regex::icase};
```
Comment (CheSema) [inline src/Server/MySQLHandler.cpp:541]: `std::regex::icase` -- Case insensitive.
you need to pass `re2::RE2::Options::set_case_sensitive(false)` to the `expr`.
As I see `case_sensitive` is `true` by default in re2.

File: src/Parsers/Kusto/KustoFunctions/KQLCastingFunctions.cpp [hunk 2]
```diff
@@ -4,9 +4,17 @@
 #include <Parsers/Kusto/KustoFunctions/KQLFunctionFactory.h>
 
 #include <format>
-#include <regex>
 #include <Poco/String.h>
 
+#ifdef __clang__
```
Comment (CheSema) [inline src/Parsers/Kusto/KustoFunctions/KQLCastingFunctions.cpp:9]: May be we should wrap that staff in some our header? It looks excessive.

Comment (rschu1ze) [reply src/Parsers/Kusto/KustoFunctions/KQLCastingFunctions.cpp:9]: Agree, I'll check.

File: src/Server/MySQLHandler.cpp (hunk 1 above)
Comment (rschu1ze) [reply src/Server/MySQLHandler.cpp:541]: Correct, thanks, done.

**Review by CheSema — APPROVED** [review]

**Comment by Alex-Cheng** [comment]
@rschu1ze  RE2 does not support `(?!re)` and `(?=re)` which I need. When I switch to std::regex I got a compilation error (it is prohibited in ClickHouse build configuration). How can I resolve the problem?

image

**Comment by rschu1ze** [comment]
`(?!re)` and `(?=re)` are negative and positive lookaround assertions.

Quoting from https://github.com/google/re2/wiki/WhyRE2: "It is also not a goal to implement all of the features offered by Perl, PCRE and other engines. As a matter of principle, RE2 does not support constructs for which only backtracking solutions are known to exist. Thus, backreferences and look-around assertions are not supported."

Perhaps you could try to use re2's non-capturing groups: `(?:re)` instead?

**Comment by Alex-Cheng** [comment]
> @Alex-Cheng `(?!re)` and `(?=re)` are negative and positive lookaround assertions.
> 
> Quoting from https://github.com/google/re2/wiki/WhyRE2: "It is also not a goal to implement all of the features offered by Perl, PCRE and other engines. As a matter of principle, RE2 does not support constructs for which only backtracking solutions are known to exist. Thus, backreferences and look-around assertions are not supported."
> 
> Perhaps you could try to use re2's non-capturing groups: `(?:re)` instead?

Thank you.
Actually it is a legacy code that uses `(?!re)`, and I have to understand the code and try to transform it to what RE2 supports. The old code from my understand is to do the task described as following:

> find the string starting with `B` and ending with `E`, with an intermediate node which is NOT `E` nor `C`.

The original regex expression is `/(>B>)((?!(>E>)|(>C>))(>[^>]*>))(>E>)/gm`
image

**Pull request merged**